lib/task_client.js

```diff
@@ -189,24 +190,42 @@ exports.initialize = function(options) {
   /**
    * Shutdown the TaskClient
    */
-  TaskClient.prototype.shutdown = function() {
+  TaskClient.prototype.shutdown = function(callback) {
     var that = this;
 
+    var shutdownFunctions = [that.notificationBus.shutdown.bind(that.notificationBus)];
+
     if (!that.isAvailable) {
       return; // Nothing to do
     }
 
     that.isAvailable = false;
     // Hard end, don't worry about shutting down
     // gracefully here...
+
+    for (var taskId in that.msgMap) {
+      if (!that.msgMap.hasOwnProperty(taskId)) {
+        continue;
+      }
+
+      cleanupMaps(that, taskId);
+    }
+
     if (that.blacklist) {
-      that.blacklist.shutdown();
+      shutdownFunctions.push(that.blacklist.shutdown.bind(that.blacklist));
     }
+
     if (that.taskLimit) {
-      that.taskLimit.shutdown();
+      shutdownFunctions.push(that.taskLimit.shutdown.bind(that.taskLimit));
     }
-    that.notificationBus.shutdown();
-    that.removeAllListeners();
+
+    async.parallel(shutdownFunctions, function(err) {
+      // Don't worry about errors
+      that.removeAllListeners();
+      if (typeof callback === 'function') {
```

Comment:
Not actually a problem but more of a question for my own understanding... in what circumstances could `callback !== function`?